From 40dcc6589f187b031bd5856ac6627dc8a4b91902 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Mon, 2 Mar 2026 00:54:49 +0800
Subject: [PATCH 58/89] mtk: mt76: mt7996: Add lpi support with sku_idx and
 enhancement

Add lpi support with sku_idx and enhancement.
1. Add sku index for lpi sku table
2. Add lpi psd limit for psd country and compensate power by fw.
3. Add lpi mode mamangement from enhancement that use 1T to transmit mgmt
from in 80 MHz.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>

Fix wrong maximum power by checkinh max_power of operating channel.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 eeprom.c             |  11 +++-
 mt76.h               |   7 ++-
 mt76_connac3_mac.h   |  12 +++++
 mt7996/init.c        |   2 +-
 mt7996/mac.c         |   5 ++
 mt7996/main.c        |   5 +-
 mt7996/mcu.c         | 119 ++++++++++++++++++++++++++++++++++++++---
 mt7996/mcu.h         |   1 +
 mt7996/mt7996.h      |   4 +-
 mt7996/mtk_debugfs.c |   3 +-
 mt7996/vendor.c      | 124 +++++++++++++++++++++++++++++++++++++++++++
 mt7996/vendor.h      |  15 ++++++
 14 files changed, 296 insertions(+), 26 deletions(-)

--- a/mt76_connac3_mac.h
+++ b/mt76_connac3_mac.h
@@ -204,6 +204,18 @@ enum tx_frag_idx {
 	MT_TX_FRAG_LAST
 };
 
+/* BW defined in FW hal_cal_flow_rom.h */
+enum {
+	FW_CDBW_20MHZ,
+	FW_CDBW_40MHZ,
+	FW_CDBW_80MHZ,
+	FW_CDBW_160MHZ,
+	FW_CDBW_320MHZ,
+	FW_CDBW_5MHZ,
+	FW_CDBW_10MHZ,
+	FW_CDBW_8080MHZ,
+};
+
 #define MT_CT_INFO_APPLY_TXD		BIT(0)
 #define MT_CT_INFO_COPY_HOST_TXD_ALL	BIT(1)
 #define MT_CT_INFO_MGMT_FRAME		BIT(2)
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -371,7 +371,7 @@ static void __mt7996_init_txpower(struct
 
 	phy->sku_limit_en = true;
 	phy->sku_path_en = true;
-	np = mt76_find_power_limits_node(&dev->mt76);
+	np = mt76_find_power_limits_node(phy->mt76);
 	for (i = 0; i < sband->n_channels; i++) {
 		struct ieee80211_channel *chan = &sband->channels[i];
 		int target_power = mt7996_eeprom_get_target_power(dev, chan);
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -883,8 +883,13 @@ void mt7996_mac_write_txwi(struct mt7996
 		}
 
 		val = FIELD_PREP(MT_TXD6_TX_RATE, idx) | MT_TXD6_FIXED_BW;
+
 		if (mcast)
 			val |= MT_TXD6_DIS_MAT;
+		if (dev->mt76.phys[band_idx]->cap.has_6ghz &&
+		    dev->mt76.lpi_bcn_enhance &&
+		    ieee80211_is_mgmt(hdr->frame_control))
+			val |= FIELD_PREP(MT_TXD6_BW, FW_CDBW_80MHZ);
 		txwi[6] |= cpu_to_le32(val);
 		txwi[3] |= cpu_to_le32(MT_TXD3_BA_DISABLE);
 	}
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -819,6 +819,7 @@ static u8
 mt7996_get_rates_table(struct mt7996_phy *phy, struct ieee80211_bss_conf *conf,
 		       bool beacon, bool mcast)
 {
+#define FR_RATE_IDX_OFDM_6M 0x004b
 	struct mt7996_dev *dev = phy->dev;
 	struct mt76_vif_link *mvif = mt76_vif_conf_link(&dev->mt76, conf->vif, conf);
 	u16 rate;
@@ -827,6 +828,9 @@ mt7996_get_rates_table(struct mt7996_phy
 	rate = mt76_connac2_mac_tx_rate_val(phy->mt76, conf, beacon, mcast);
 
 	if (beacon) {
+		if (dev->mt76.lpi_bcn_enhance)
+			rate = FR_RATE_IDX_OFDM_6M;
+
 		/* odd index for driver, even index for firmware */
 		idx = MT7996_BEACON_RATES_TBL + 2 * phy->mt76->band_idx;
 		if (phy->beacon_rate != rate)
@@ -1008,7 +1012,7 @@ mt7996_link_info_changed(struct ieee8021
 	if (changed & BSS_CHANGED_TXPOWER &&
 	    info->txpower != phy->txpower) {
 		phy->txpower = info->txpower;
-		mt7996_mcu_set_txpower_sku(phy);
+		mt7996_mcu_set_txpower_sku(phy, info->txpower);
 	}
 
 out:
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -5524,7 +5524,7 @@ int mt7996_mcu_set_fixed_rate_table(stru
 
 	if (beacon) {
 		req.spe_idx_sel = SPE_IXD_SELECT_TXD;
-		req.spe_idx = 24 + band_idx;
+		req.spe_idx = dev->mt76.mgmt_pwr_enhance ? 0 : 24 + band_idx;
 		phy->beacon_rate = rate_idx;
 	} else {
 		req.spe_idx_sel = SPE_IXD_SELECT_BMC_WTBL;
@@ -5838,6 +5838,22 @@ mt7996_update_max_txpower_cur(struct mt7
 		mphy->txpower_cur = e2p_power_limit;
 }
 
+static bool mt7996_is_psd_country(char *country)
+{
+	char psd_country_list[][3] = {"US", "KR", "BR", "CL", "MY", ""};
+	int i;
+
+	if (strlen(country) != 2)
+		return 0;
+
+	for (i = 0; psd_country_list[i][0] != '\0'; i++) {
+		if (!strncmp(country, psd_country_list[i], 2))
+			return 1;
+	}
+
+	return 0;
+}
+
 static s8 mt7996_get_power_bound(struct mt7996_phy *phy, s8 txpower)
 {
        struct mt76_phy *mphy = phy->mt76;
@@ -5849,7 +5865,8 @@ static s8 mt7996_get_power_bound(struct
        return txpower;
 }
 
-int mt7996_mcu_set_txpower_sku(struct mt7996_phy *phy)
+int mt7996_mcu_set_txpower_sku(struct mt7996_phy *phy,
+			       int txpower_setting)
 {
 #define TX_POWER_LIMIT_TABLE_RATE	0
 #define TX_POWER_LIMIT_TABLE_PATH	1
@@ -5874,7 +5891,9 @@ int mt7996_mcu_set_txpower_sku(struct mt
 	struct sk_buff *skb;
 	int i, ret, txpower_limit;
 
-	txpower_limit = mt7996_get_power_bound(phy, phy->txpower);
+	if (txpower_setting == INT_MIN || txpower_setting > 127)
+		txpower_setting = 127;
+	txpower_limit = mt7996_get_power_bound(phy, txpower_setting);
 
 	if (phy->sku_limit_en) {
 		la = kzalloc(sizeof(*la), GFP_KERNEL);
@@ -5897,7 +5916,37 @@ int mt7996_mcu_set_txpower_sku(struct mt
 	}
 
 	skb_put_data(skb, &req, sizeof(req));
-	skb_put_data(skb, &la->cck, 12); /* CCK + OFDM */
+	skb_put_data(skb, &la->cck, 4);
+
+	/* FW would compensate for PSD countries
+	 * driver doesn't need to do it
+	 */
+	if (phy->mt76->cap.has_6ghz && mphy->dev->lpi_psd &&
+	    !mt7996_is_psd_country(dev->mt76.alpha2)) {
+		switch (mphy->chandef.width) {
+		case NL80211_CHAN_WIDTH_20:
+			skb_put_data(skb, &la->eht[3], 8);
+			break;
+		case NL80211_CHAN_WIDTH_40:
+			skb_put_data(skb, &la->eht[4], 8);
+			break;
+		case NL80211_CHAN_WIDTH_80:
+			skb_put_data(skb, &la->eht[5], 8);
+			break;
+		case NL80211_CHAN_WIDTH_160:
+			skb_put_data(skb, &la->eht[6], 8);
+			break;
+		case NL80211_CHAN_WIDTH_320:
+			skb_put_data(skb, &la->eht[7], 8);
+			break;
+		default:
+			skb_put_data(skb, &la->ofdm, sizeof(la->ofdm));
+			break;
+		}
+	} else {
+		skb_put_data(skb, &la->ofdm, 8);
+	}
+
 	skb_put_data(skb, &la->mcs[0], 8); /* HT20 */
 	skb_put_data(skb, &la->mcs[1], 9); /* HT40 */
 
@@ -5936,8 +5985,41 @@ int mt7996_mcu_set_txpower_sku(struct mt
 
 	skb_put_data(skb, &req, sizeof(req));
 	skb_put_data(skb, &la->path.cck, 5);
-	skb_put_data(skb, &la->path.ofdm, 5);
-	skb_put_data(skb, &la->path.ofdm_bf, 4);
+
+	/* FW would NOT compensate in the case of BF backoff table
+	 * driver needs to compensate for LPI PSD
+	 */
+	if (phy->mt76->cap.has_6ghz && mphy->dev->lpi_psd) {
+		switch (mphy->chandef.width) {
+		case NL80211_CHAN_WIDTH_20:
+			skb_put_data(skb, &la->path.ru[5], 5);
+			skb_put_data(skb, &la->path.ru_bf[5], 4);
+			break;
+		case NL80211_CHAN_WIDTH_40:
+			skb_put_data(skb, &la->path.ru[6], 5);
+			skb_put_data(skb, &la->path.ru_bf[6], 4);
+			break;
+		case NL80211_CHAN_WIDTH_80:
+			skb_put_data(skb, &la->path.ru[8], 5);
+			skb_put_data(skb, &la->path.ru_bf[8], 4);
+			break;
+		case NL80211_CHAN_WIDTH_160:
+			skb_put_data(skb, &la->path.ru[11], 5);
+			skb_put_data(skb, &la->path.ru_bf[11], 4);
+			break;
+		case NL80211_CHAN_WIDTH_320:
+			skb_put_data(skb, &la->path.ru[15], 5);
+			skb_put_data(skb, &la->path.ru_bf[15], 4);
+			break;
+		default:
+			skb_put_data(skb, &la->path.ofdm, 5);
+			skb_put_data(skb, &la->path.ofdm_bf, 4);
+			break;
+		}
+	} else {
+		skb_put_data(skb, &la->path.ofdm, 5);
+		skb_put_data(skb, &la->path.ofdm_bf, 4);
+	}
 
 	for (i = 0; i < 32; i++) {
 		bool bf = i % 2;
@@ -5954,6 +6036,31 @@ out:
 	return ret;
 }
 
+int mt7996_mcu_set_lpi_psd(struct mt7996_phy *phy, u8 enable)
+{
+	struct mt7996_dev *dev = phy->dev;
+
+	struct {
+		u8 band_idx;
+		u8 _rsv[3];
+
+		__le16 tag;
+		__le16 len;
+		u8 lpi_enable;
+		u8 psd_limit;
+		u8 _rsv2[2];
+	} __packed req = {
+		.band_idx = phy->mt76->band_idx,
+		.tag = cpu_to_le16(UNI_BAND_CONFIG_LPI_CTRL),
+		.len = cpu_to_le16(sizeof(req) - 4),
+		.lpi_enable = enable,
+		.psd_limit = enable ? mt7996_is_psd_country(dev->mt76.alpha2) : 0,
+	};
+
+	return mt76_mcu_send_msg(&phy->dev->mt76, MCU_WM_UNI_CMD(BAND_CONFIG),
+				 &req, sizeof(req), false);
+}
+
 int mt7996_mcu_cp_support(struct mt7996_dev *dev, u8 mode)
 {
 	__le32 cp_mode;
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -979,6 +979,7 @@ enum {
 	UNI_BAND_CONFIG_EDCCA_THRESHOLD = 0x06,
 	UNI_BAND_CONFIG_RTS_THRESHOLD = 0x08,
 	UNI_BAND_CONFIG_MAC_ENABLE_CTRL = 0x0c,
+	UNI_BAND_CONFIG_LPI_CTRL = 0x0d,
 };
 
 enum {
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -862,7 +862,8 @@ int mt7996_mcu_get_chan_mib_info(struct
 int mt7996_mcu_get_temperature(struct mt7996_phy *phy);
 int mt7996_mcu_set_thermal_throttling(struct mt7996_phy *phy, u8 state);
 int mt7996_mcu_set_thermal_protect(struct mt7996_phy *phy, bool enable);
-int mt7996_mcu_set_txpower_sku(struct mt7996_phy *phy);
+int mt7996_mcu_set_txpower_sku(struct mt7996_phy *phy, int txpower_setting);
+int mt7996_mcu_set_lpi_psd(struct mt7996_phy *phy, u8 enable);
 int mt7996_mcu_rdd_resume_tx(struct mt7996_phy *phy);
 int mt7996_mcu_rdd_cmd(struct mt7996_dev *dev, int cmd, u8 rdd_idx, u8 val);
 int mt7996_mcu_rdd_background_enable(struct mt7996_phy *phy,
--- a/mt7996/mtk_debugfs.c
+++ b/mt7996/mtk_debugfs.c
@@ -2437,7 +2437,7 @@ mt7996_get_txpower_info(struct file *fil
 	len += scnprintf(buf + len, size - len,
 			 "    Theraml Compensation Value: %d\n",
 			 basic_info->thermal_compensate_value);
-	np = mt76_find_power_limits_node(phy->mt76->dev);
+	np = mt76_find_power_limits_node(phy->mt76);
 	len += scnprintf(buf + len, size - len,
 			 "    RegDB:  %s\n",
 			 !np ? "enable" : "disable");
@@ -2894,6 +2894,7 @@ int mt7996_mtk_init_debugfs(struct mt799
 				    mt7996_wtbl_read);
 
 	debugfs_create_u8("sku_disable", 0600, dir, &dev->dbg.sku_disable);
+	debugfs_create_bool("mgmt_pwr_enhance", 0600, dir, &dev->mt76.mgmt_pwr_enhance);
 
 	for (i = 0; i < hw->wiphy->n_radio; i++) {
 		char buf[32];
--- a/mt7996/vendor.c
+++ b/mt7996/vendor.c
@@ -35,6 +35,14 @@ edcca_ctrl_policy[NUM_MTK_VENDOR_ATTRS_E
 	[MTK_VENDOR_ATTR_EDCCA_CTRL_RADIO_IDX] = { .type = NLA_U8 },
 };
 
+static struct nla_policy
+txpower_ctrl_policy[NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL] = {
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID] = { .type = NLA_U8 },
+};
+
 static const struct nla_policy
 edcca_dump_policy[NUM_MTK_VENDOR_ATTRS_EDCCA_DUMP] = {
 	[MTK_VENDOR_ATTR_EDCCA_DUMP_MODE] = { .type = NLA_U8 },
@@ -321,6 +329,111 @@ error:
 	return -EINVAL;
 }
 
+static int mt7996_vendor_txpower_ctrl(struct wiphy *wiphy,
+				      struct wireless_dev *wdev,
+				      const void *data,
+				      int data_len)
+{
+#define FR_RATE_IDX_OFDM_6M 0x004b
+	struct mt7996_dev *dev;
+	struct mt7996_phy *phy;
+	struct mt76_phy *mphy;
+	struct ieee80211_vif *vif = wdev_to_ieee80211_vif(wdev);
+	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
+	struct mt7996_vif_link *mconf;
+	struct nlattr *tb[NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL];
+	struct mt76_power_limits la = {};
+	struct mt76_power_path_limits la_path = {};
+	u8 orig_sku_idx = mphy->sku_idx;
+	int err, current_txpower, delta;
+	u8 val, link_id = 0, idx;
+	s8 new_power_limit = 127;
+
+	err = nla_parse(tb, MTK_VENDOR_ATTR_TXPOWER_CTRL_MAX, data, data_len,
+			txpower_ctrl_policy, NULL);
+	if (err)
+		return err;
+
+
+	if (ieee80211_vif_is_mld(vif) && tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID]) {
+		link_id = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID]);
+
+		if (link_id >= IEEE80211_LINK_UNSPECIFIED)
+			return -EINVAL;
+	}
+
+	rcu_read_lock();
+	mconf = (struct mt7996_vif_link *)rcu_dereference(mvif->mt76.link[link_id]);
+	if (!mconf || !mconf->phy) {
+		rcu_read_unlock();
+		return -EINVAL;
+	}
+
+	phy = mconf->phy;
+	rcu_read_unlock();
+
+	mphy = phy->mt76;
+
+	if (mphy->cap.has_6ghz &&
+	    tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD]) {
+		val = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD]);
+		mphy->dev->lpi_psd = val;
+
+		err = mt7996_mcu_set_lpi_psd(phy, val);
+		if (err)
+			return err;
+	}
+
+	if (tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX]) {
+		struct ieee80211_channel *chan = mphy->main_chandef.chan;
+
+		mphy->sku_idx = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX]);
+
+		if (mt76_find_power_limits_node(mphy) == NULL)
+			mphy->sku_idx = 0;
+
+		if (orig_sku_idx != mphy->sku_idx)
+			mt7996_init_txpower(phy);
+		new_power_limit = chan->max_power;
+
+		phy->sku_limit_en = true;
+		phy->sku_path_en = true;
+		mt76_get_rate_power_limits(mphy, chan, la, 127);
+		if (!la_path.ofdm[0])
+			phy->sku_path_en = false;
+
+		dev = phy->dev;
+		err = mt7996_mcu_set_tx_power_ctrl(phy, UNI_TXPOWER_SKU_POWER_LIMIT_CTRL,
+						   dev->dbg.sku_disable ? 0 : phy->sku_limit_en);
+		if (err)
+			return err;
+		err = mt7996_mcu_set_tx_power_ctrl(phy, UNI_TXPOWER_BACKOFF_POWER_LIMIT_CTRL,
+						   dev->dbg.sku_disable ? 0 : phy->sku_path_en);
+		if (err)
+			return err;
+	}
+
+	if (mphy->cap.has_6ghz &&
+	    tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE]) {
+		val = nla_get_u8(tb[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE]);
+		mphy->dev->lpi_bcn_enhance = val;
+		idx = MT7996_BEACON_RATES_TBL + 2 * phy->mt76->band_idx;
+
+		err = mt7996_mcu_set_fixed_rate_table(phy, idx, FR_RATE_IDX_OFDM_6M, true);
+		if (err)
+			return err;
+	}
+
+	current_txpower = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
+	new_power_limit = min_t(s8, current_txpower, new_power_limit);
+
+	err = mt7996_mcu_set_txpower_sku(phy, new_power_limit);
+	if (err)
+		return err;
+
+	return 0;
+}
+
 static const struct wiphy_vendor_command mt7996_vendor_commands[] = {
 	{
 		.info = {
@@ -359,6 +472,17 @@ static const struct wiphy_vendor_command
 	},
 	{
 		.info = {
+			.vendor_id = MTK_NL80211_VENDOR_ID,
+			.subcmd = MTK_NL80211_VENDOR_SUBCMD_TXPOWER_CTRL,
+		},
+		.flags = WIPHY_VENDOR_CMD_NEED_NETDEV |
+			 WIPHY_VENDOR_CMD_NEED_RUNNING,
+		.doit = mt7996_vendor_txpower_ctrl,
+		.policy = txpower_ctrl_policy,
+		.maxattr = MTK_VENDOR_ATTR_TXPOWER_CTRL_MAX,
+	},
+	{
+		.info = {
 			.vendor_id = MTK_NL80211_VENDOR_ID,
 			.subcmd = MTK_NL80211_VENDOR_SUBCMD_PP_CTRL,
 		},
--- a/mt7996/vendor.h
+++ b/mt7996/vendor.h
@@ -8,6 +8,21 @@ enum mtk_nl80211_vendor_subcmds {
 	MTK_NL80211_VENDOR_SUBCMD_EDCCA_CTRL = 0xc7,
 	MTK_NL80211_VENDOR_SUBCMD_BSS_COLOR_CTRL = 0xca,
 	MTK_NL80211_VENDOR_SUBCMD_PP_CTRL = 0xcc,
+	MTK_NL80211_VENDOR_SUBCMD_TXPOWER_CTRL = 0xce,
+};
+
+enum mtk_vendor_attr_txpower_ctrl {
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_UNSPEC,
+
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID,
+
+	/* keep last */
+	NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_MAX =
+		NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL - 1
 };
 
 enum mtk_vendor_attr_edcca_ctrl {
